Server/JokesOnYou.Web.Api/Services/AuthService.cs

```diff
@@ -1,22 +1,60 @@
 ﻿using JokesOnYou.Web.Api.DTOs;
+using JokesOnYou.Web.Api.Exceptions;
 using JokesOnYou.Web.Api.Models;
+using JokesOnYou.Web.Api.Repositories.Interfaces;
+using JokesOnYou.Web.Api.Services.Interfaces;
+using Microsoft.AspNetCore.Identity;
 using System;
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
 
 namespace JokesOnYou.Web.Api.Services
 {
-    public class AuthService
+    public class AuthService : IAuthService
     {
-        User Login(UserLoginDTO userLoginDTO)
+        private readonly SignInManager<User> _signInManager;
+        private readonly IUserRepository _userRepo;
+        private readonly ITokenService _tokenService;
+
+        public AuthService(SignInManager<User> signInManager, IUserRepository userRepo, ITokenService tokenService)
         {
-            throw new NotImplementedException();
+            _signInManager = signInManager;
+            _userRepo = userRepo;
+            _tokenService = tokenService;
+        }
+
+        public async Task<UserReplyDTO> LoginAsync(UserLoginDTO userLoginDTO)
+        {
+            var user = await _userRepo.GetUserByEmail(userLoginDTO.Email);
+
+            var result = await _signInManager.CheckPasswordSignInAsync(user, userLoginDTO.Password, false);
+
+            //var result = await _signInManager.PasswordSignInAsync(userLoginDTO.Email, userLoginDTO.Password, false, false);
+
+            if (result.Succeeded)
+            {
+                //TODO need to assign user a role!
+                UserReplyDTO userReplyDTO = new UserReplyDTO() { Email = user.Email, Id = user.Id, UserName = user.UserName, Role = "Registered" };
+
+                string token = _tokenService.GetToken(user);
```

Comment:
for some more decoupling, maybe our token creator should just recieve what it needs, like the roles and username

Reply:
I'm not so sure about this one.
We could but would mean more work sometimes.
```
Every time we want to add anything to the GetToken we have to change this method and the GetToken.
If we add more places where GetToken gets called, they all have to change their parameters when GetToken changes.
```
```
Now if we add stuff to the Token we don't need to change this nor the parameters.
```
So i think this is better. But please do convince me!